s3store/s3store_test.go

```diff
@@ -24,6 +24,138 @@ var _ tusd.TerminaterDataStore = s3store.S3Store{}
+	}
+	var HighestApplicablePartSize int64 = MaxObjectSize / MaxMultipartParts
+	if MaxObjectSize%MaxMultipartParts > 0 {
+		HighestApplicablePartSize++
+	}
+	var RemainderWithHighestApplicablePartSize int64 = MaxObjectSize % HighestApplicablePartSize
+
+	// some of these tests are actually duplicates, as they specify the same size
+	// in bytes - two ways to describe the same thing. That is wanted, in order
+	// to provide a full picture from any angle.
+	testcases := []int64{
+		1,
+		MinPartSize - 1,
+		MinPartSize,
+		MinPartSize + 1,
+
+		MinPartSize*(MaxMultipartParts-1) - 1,
+		MinPartSize * (MaxMultipartParts - 1),
+		MinPartSize*(MaxMultipartParts-1) + 1,
+
+		MinPartSize*MaxMultipartParts - 1,
+		MinPartSize * MaxMultipartParts,
+		MinPartSize*MaxMultipartParts + 1,
+
+		MinPartSize*(MaxMultipartParts+1) - 1,
+		MinPartSize * (MaxMultipartParts + 1),
+		MinPartSize*(MaxMultipartParts+1) + 1,
+
+		(HighestApplicablePartSize-1)*MaxMultipartParts - 1,
+		(HighestApplicablePartSize - 1) * MaxMultipartParts,
+		(HighestApplicablePartSize-1)*MaxMultipartParts + 1,
+
+		HighestApplicablePartSize*(MaxMultipartParts-1) - 1,
+		HighestApplicablePartSize * (MaxMultipartParts - 1),
+		HighestApplicablePartSize*(MaxMultipartParts-1) + 1,
+
+		HighestApplicablePartSize*(MaxMultipartParts-1) + RemainderWithHighestApplicablePartSize - 1,
+		HighestApplicablePartSize*(MaxMultipartParts-1) + RemainderWithHighestApplicablePartSize,
+		HighestApplicablePartSize*(MaxMultipartParts-1) + RemainderWithHighestApplicablePartSize + 1,
+
+		MaxObjectSize - 1,
+		MaxObjectSize,
+		MaxObjectSize + 1,
+
+		(MaxObjectSize/MaxMultipartParts)*(MaxMultipartParts-1) - 1,
+		(MaxObjectSize / MaxMultipartParts) * (MaxMultipartParts - 1),
+		(MaxObjectSize/MaxMultipartParts)*(MaxMultipartParts-1) + 1,
+
+		MaxPartSize*(MaxMultipartParts-1) - 1,
+		MaxPartSize * (MaxMultipartParts - 1),
+		MaxPartSize*(MaxMultipartParts-1) + 1,
+
+		MaxPartSize*MaxMultipartParts - 1,
+		MaxPartSize * MaxMultipartParts,
+		MaxPartSize*MaxMultipartParts + 1,
+	}
+
+	for index, size := range testcases {
+		err = ""
+		optimalPartSize, calcError := store.CalcOptimalPartSize(size)
+		if size > MaxObjectSize && calcError == nil {
+			err += fmt.Sprintf("Testcase #%v size %v: size exceeds MaxObjectSize %v but no error returned\n", index, size, MaxObjectSize)
+		}
+		if debug && optimalPartSize == 0 {
+			fmt.Printf("Testcase #%v size %v: size exceeds MaxObjectSize %v\n", index, size, MaxObjectSize)
+		}
+		if optimalPartSize > 0 {
+			equalparts = size / optimalPartSize
+			lastpartsize = size % optimalPartSize
+			if optimalPartSize < MinPartSize {
+				err += fmt.Sprintf("Testcase #%v size %v, %v parts of size %v, lastpart %v: optimalPartSize < MinPartSize %v\n", index, size, equalparts, optimalPartSize, lastpartsize, MinPartSize)
```

Comment:
Any reason why you are not using t.Errorf here and in the following if statement?

Reply:
No. As said, with error handling, I am hoping for your advice. My Golang experience is limited and I did not read up on everything yet. I was generating strings and only using t.Errorf at the bottom, to make sure, I see all errors not just the first one. Don't know, if this is necessary.

Reply:
That's why I asked :) FYI, t.Error will continue execution after it has been invoked while t.Fatal will immediately stop the test case (see https://golang.org/pkg/testing/#B.Errorf). For example, I use t.Fatal if a database connection could not be made and further testing is useless. On the other hand, t.Error is used for asserting that the result has specific properties.